Source/qol/stash.cpp

```diff
@@ -494,53 +449,64 @@ bool UseStashItem(uint16_t c)
 
 	UseItem(MyPlayerId, item->_iMiscId, item->_iSpell);
 
-	if (Stash.stashList[c]._iMiscId == IMISC_MAPOFDOOM)
+	if (Stash.stashList[itemIndex]._iMiscId == IMISC_MAPOFDOOM)
 		return true;
-	if (Stash.stashList[c]._iMiscId == IMISC_NOTE) {
+	if (Stash.stashList[itemIndex]._iMiscId == IMISC_NOTE) {
 		InitQTextMsg(TEXT_BOOK9);
 		CloseInventory();
 		return true;
 	}
-	Stash.RemoveStashItem(c);
+	Stash.RemoveStashItem(item, itemPosition, itemIndex);
 
 	return true;
 }
 
-void StashStruct::RemoveStashItem(StashStruct::StashCell iv)
+void StashStruct::RemoveStashItem(Item *item, Point itemPosition, StashCell itemIndex)
 {
-	// Iterate through stashGrid and remove every reference to item
-	for (auto &row : Stash.GetCurrentGrid()) {
-		for (StashStruct::StashCell &itemId : row) {
-			if (itemId - 1 == iv) {
-				itemId = 0;
-			}
-		}
-	}
-
-	if (stashList.empty()) {
+	if (item == nullptr) {
 		return;
 	}
 
-	// If the item at the end of stash array isn't the one we removed, we need to swap its position in the array with the removed item
+	// get item's size to figure slot changes
+	const Size itemSize = GetInventorySize(*item);
+
+	// reset claimed slot IDs to zero
+	UpdateStashGrid(Stash.GetPage(), itemPosition, itemSize);
+
+	// not the last item, swap with last item to keep IDs contiguous
 	StashStruct::StashCell lastItemIndex = static_cast<StashStruct::StashCell>(stashList.size() - 1);
-	if (lastItemIndex != iv) {
-		stashList[iv] = stashList[lastItemIndex];
-
-		for (auto &pair : Stash.stashGrids) {
-			auto &grid = pair.second;
-			for (auto &row : grid) {
-				for (StashStruct::StashCell &itemId : row) {
-					if (itemId == lastItemIndex + 1) {
-						itemId = iv + 1;
-					}
-				}
-			}
-		}
+	if (lastItemIndex != itemIndex) {
+		Point swappedItemPosition = stashList[lastItemIndex].position;
+		const Size swappedItemSize = GetInventorySize(stashList[lastItemIndex]);
+		unsigned swappedItemPage = GetPageByItemId(lastItemIndex);
+		std::swap(stashList[itemIndex], stashList[lastItemIndex]);
+
+		UpdateStashGrid(swappedItemPage, swappedItemPosition, swappedItemSize, itemIndex + 1);
 	}
+
+	// drop last item (our item)
 	stashList.pop_back();
+
+	// flag for update when sfile_write_stash() is called
 	Stash.dirty = true;
 }
 
+unsigned StashStruct::GetPageByItemId(StashCell itemId)
+{
+	if (itemId < stashList.size()) {
+		Point itemPosition = stashList[itemId].position;
+		unsigned currentPageIndex = 0;
+		while (currentPageIndex <= LastStashPage) {
+			if (stashGrids[currentPageIndex][itemPosition.x][itemPosition.y] == itemId + 1) {

```

Comment:
This causes allocations if a player has skipped pages when navigating the stash. You can use `std::map::iterator` based loops instead.

```suggestion
		for (const auto &[currentPageIndex, stashGrid] : stashGrids) { // C++17, probably need to use the long form for GCC6 support
			if (stashGrid[itemPosition.x][itemPosition.y] == itemId + 1) {
```

Reply:
The intent here is to find an itemId that may not necessarily be on the current page. At the time, it seemed to be less complex to start from 0 than to start at the current page then wrap around.

Reply:
That's fine, `stashGrid` is a `std::map` which orders keys by natural ordering. The range based loop will start from the first already allocated page and continue through all pages that already exist. Using `operator[]` to access a key that doesn't exist will end up inserting a blank page if someone skips from page 5 to 10.

As an example, lets say I order things in the stash so I have two pages of potions starting from page 1 (key 0), three pages of armour from page 10, and two pages of weapons from page 20. I then call this function with an index that doesn't exist (so the early return never happens). The range based loop will iterate over the pairs with keys [0, 1, 9, 10, 11, 19, 20] in that order (assuming the player hasn't flicked through the stash since last loading the game). The index based loop will iterate over [0..99], allocating 93 new empty grids.

Reply:
Excellent explanation and I did change this just now to move away form range based iteration.